packages/client/src/runtime/DataProxyEngine.ts

```diff
@@ -0,0 +1,282 @@
+
+  request<T>(query: string, headers: Record<string, string>, attempt = 0) {
+    this.logEmitter.emit('query', { query })
+
+    return this.requestInternal<T>({ query, variables: {} }, headers, attempt)
+  }
+
+  async requestBatch<T>(
+    queries: string[],
+    headers: Record<string, string>,
+    isTransaction = false,
+    attempt = 0,
+  ) {
+    this.logEmitter.emit('query', {
+      query: `Batch${isTransaction ? ' in transaction' : ''} (${
+        queries.length
+      }):\n${queries.join('\n')}`,
+    })
+
+    const body = {
+      batch: queries.map((query) => ({ query, variables: {} })),
+      transaction: isTransaction,
+    }
+
+    const { batchResult } = await this.requestInternal<T>(
+      body,
+      headers,
+      attempt,
+    )
+
+    return batchResult
+  }
+
+  private async requestInternal<T>(
+    body: Record<string, any>,
+    headers: Record<string, string>,
+    attempt: number,
+  ) {
+    try {
+      this.logEmitter.emit('info', {
+        message: `Calling ${this.url('graphql')} (n=${attempt})`,
+      })
+
+      const res = await fetch(this.url('graphql'), {
+        method: 'POST',
+        headers: { ...headers, ...this.headers },
+        body: JSON.stringify(body),
+      })
+
+      // 404 on the GraphQL endpoint may mean that the schema
+      // was not uploaded yet.
+      if (res.status === 404) {
+        await this.uploadSchema()
+        throw new Error('Schema (re)uploaded')
+      }
+
+      if (!res.ok) {
+        throw new Error('GraphQL request failed')
+      }
+
+      return JSON.parse(await res.text())
+    } catch (err) {
+      if (attempt >= MAX_RETRIES) {
+        this.logEmitter.emit('error', {
+          message: `Failed to query: ${err.message}`,
+        })
+        throw err
+      } else {
+        const delay = await backOff(attempt)
+        this.logEmitter.emit('warn', { message: `Retrying after ${delay}ms` })
+        return this.requestInternal<T>(body, headers, attempt + 1)
+      }
+    }
+  }
+
+  // TODO: figure out how to support transactions
+  transaction(): Promise<any> {
+    throw new Error('Transactions are currently not supported in Data Proxy')
+  }
+}
+
+// Regex + prismafile feels fragile...
+// TODO: build a minimal single-purpose (URL extraction)
+// and well-tested TS schema parser
+function extractHostAndApiKey(schemaText: string) {
+  // Extract the datasource block so we feed prismafile with minimal input
+  const strippedToDatasource = /datasource[ \t]+[^\s]+[ \t]+\{[^}]+}/.exec(
+    schemaText,
+  )
+  if (!strippedToDatasource) {
+    throw new Error('Could not find a valid datasource block in the schema')
+  }
+
+  const datasourceText = strippedToDatasource[0]
+  if (typeof datasourceText !== 'string' || datasourceText.length < 1) {
+    throw new Error('Could not find a valid datasource block in the schema')
+  }
+
+  // Use prismafile to determine the URL
+  let url = ''
+  try {
+    const schemaAst = prismafile.parse(datasourceText)
```

Comment:
I really didn't expect us to start using this officially. Just a note that it's not 100% feature compatible with our current schema.prisma. I think it's missing @@index and a few other things 😅

Maybe instead of using this in-full, we could generate a parser that just specializes in extracting datasources.

E.g. https://github.com/prisma/prismafile/blob/master/src/parser/index.pegjs#L51

Reply:
Could this parsing be tackled by the engines any time soon?

Reply:
Problem with doing this in Engines is that it still requires an Engine being present. With the binary and library variants that means downloading a platform dependent file (including it in the generated Client). With WASM this could be a Node package, but we do not have that yet and are making first experiences with `prisma-fmt` only. (This could of course be a good reason to prioritize this work)

Reply:
Hmm, I was thinking about using the engines during the generation step to help embedding the URL or the name of the ENV var into the generated code.

Reply:
I actually think that staying binary-free here makes a lot of sense. I investigated writing a small focused parser for getting `datasource.url` value (or env var name) but `prismafile` does its job just fine. Note that we don't need full parsing, we're just interested in `datasource.url` (hence stripping everything but the datasource block with a regex).

Reply:
Via https://github.com/prisma/prisma/pull/9422, the schema is inlined into the generated client during the generation step. That is so that it can be embedded into browser environments (without fs). That also means that it has been pre-validated by the engines. For that reason it is now extracted via a regexp. (cc @steebchen)

Reply:
Please see this thread: https://prisma-company.slack.com/archives/C02FZENHD4N/p1633515267034400

Reply:
I'm not sure if I follow - how would you get the DMMF w/o the binary? The one that I was getting in the generated clients wouldn't have the URL nor the env var now. What we did here was: trim string to relevant part with a relatively simple readable regex and then parse the datasource (applying rule: parsing > regular expression).

I'm still pretty uncomfortable with the clients, I'd appreciate more detailed explanation for my own understanding.